### ISSUE TITLE
TLS proxy to AWS Keyspaces

### ISSUE THREAD

**Issue by vittoriocanilli** [shotover/shotover-proxy#708]
## Is your feature request related to a problem? Please describe.
Hello,

I have a PHP third-party application, which needs to access a Cassandra keyspace and it does not support TLS connections; I have set up a Cassandra keyspace in AWS Keyspaces, which requires TLS connections: therefore I need to set up a kind of proxy.

After that I have followed these [instructions](https://docs.aws.amazon.com/keyspaces/latest/devguide/programmatic.cqlsh.html), I managed successfully to connect to my Cassandra keyspace on AWS Keyspaces locally via `cqlsh`.

While I was looking for a solution, I have found Shotover proxy, but I am not sure if it is the right one for my issue. Has somebody else tested this proxy solution with AWS Keyspaces already? Thanks in advance.

## Describe the solution you'd like
The PHP third-party application is running in a Kubernetes container, which has a [ProxySQL](https://hub.docker.com/r/proxysql/proxysql) side container: ProxySQL accepts MySQL insecure MySQL connections from localhost (the PHP application, not able to support secure MySQL connections) and forwards them as secure MySQL connections to the database in AWS RDS.

With Shotover I would like to implement a proxy in a new side container, which does for the Cassandra connections what the ProxySQL side container does for the MySQL connections.

## Describe alternatives you've considered
I have searched for similar proxy solutions, but I did not find any yet. The support of the PHP third-party application could not disclose when and if any support to secure Cassandra connections will be given. AWS support did not provide a solution yet.

## Additional context
It would be nice to have an example with a Dockerfile where the necessary environment variables are passed in order to have such a proxy running.

**Comment by benbromhead** [comment]
Thanks for the feedback @vittoriocanilli - you _should_ be able to acheive what you are looking for here be using this example configuration https://github.com/shotover/shotover-proxy/blob/main/shotover-proxy/example-configs/cassandra-tls/topology.yaml

You will just need to remove the tls config from the source to turn off TLS:


```yaml
...
sources:
  cassandra_prod:
    Cassandra:
      listen_addr: "127.0.0.1:9043"
...

```

We publish a docker image of shotover here - https://hub.docker.com/r/shotover/shotover-proxy/tags 

Currently the docker image doesn't support environment variable style config, so you will need to bind mount a config volume or if you are using something like kubernetes use a config map to inject the topology.yaml file in. You'll also need to provide the AWS Keyspaces certificates (you can grab them using the documentation for keyspaces here in the first few steps of the java driver setup https://docs.aws.amazon.com/keyspaces/latest/devguide/using_java_driver.html) 

You'll use the keyspaces endpoint as well `cassandra.us-east-2.amazonaws.com`

**Comment by vittoriocanilli** [comment]
Thanks for the reply @benbromhead - I have created a Docker file as follows:

```
FROM shotover/shotover-proxy:v0.1.4

COPY files/cassandra_proxy/config.yaml /config/
COPY files/cassandra_proxy/topology.yaml /config/

ADD https://certs.secureserver.net/repository/sf-class2-root.crt /config/

#ENTRYPOINT ["./shotover-proxy"]
```

The certificate is taken from the [AWS documentation](https://docs.aws.amazon.com/keyspaces/latest/devguide/programmatic.cqlsh.html).

This is the _config.yaml_ file: 
```
---
main_log_level: "info,shotover_proxy=info"
observability_interface: "0.0.0.0:9001"
```
and this is the _topology.yaml_ file: 
```
---
sources:
  cassandra_prod:
    Cassandra:
      listen_addr: "127.0.0.1:9042"
chain_config:
  main_chain:
    - CassandraSinkSingle:
        remote_address: "cassandra.eu-west-1.amazonaws.com:9142"
        tls:
          certificate_path: "/config/sf-class2-root.crt"
source_to_chain_mapping:
  cassandra_prod: main_chain
```

I manage to build the container successfully, but when I try to run it, I get this error:
```

Error: chain_config.main_chain[0].CassandraSinkSingle.tls: missing field `certificate_authority_path` at line 12 column 27
```

So I edited the _topology.yaml_ file to use the certificate even as certificate authority:
```
---
sources:
  cassandra_prod:
    Cassandra:
      listen_addr: "127.0.0.1:9042"
chain_config:
  main_chain:
    - CassandraSinkSingle:
        remote_address: "cassandra.eu-west-1.amazonaws.com:9142"
        tls:
          certificate_authority_path: /config/sf-class2-root.crt"
          certificate_path: "/config/sf-class2-root.crt"
source_to_chain_mapping:
  cassandra_prod: main_chain
```

but then in run time I get this error message instead:
```
Error: chain_config.main_chain[0].CassandraSinkSingle.tls: missing field `private_key_path` at line 11 column 37
```

And if I keep just the certificate authority, I get this other error instead:
```
Error: chain_config.main_chain[0].CassandraSinkSingle.tls: missing field `certificate_path` at line 11 column 37
```

So it looks like Shotover needs all 3 of them: _certificate_authority_path_, _certificate_path_ and _private_key_path_. As I have only a certificate and no key, how can I proceed in this case? Thanks in advance.

**Comment by benbromhead** [comment]
@conorbros does the C* single sink require a private key to auth ourselves to upstream?

**Comment by rukai** [comment]
We have released shotover v0.1.5 https://github.com/shotover/shotover-proxy/releases/tag/v0.1.5
This new release makes certificate_path and private_key_path optional in sink transforms.
Can you try it out and let us know how it goes?

**Comment by vittoriocanilli** [comment]
Thanks for the update @rukai 

I have updated my _Dockerfile_ with `FROM shotover/shotover-proxy:v0.1.5` and my _topology.yaml_ file only has `certificate_authority_path: "/config/sf-class2-root.crt"`. I can now run the container in background forwarding the port 9042 without errors.

However, when I try to log in to AWS Keyspaces with `cqlsh` I get the following error:
```
Connection error: ('Unable to connect to any servers', {'127.0.0.1:9042': ConnectionShutdown('Connection to 127.0.0.1:9042 was closed')})
```

I have checked the connection locally on the port 9042 and this is the result:
```
% telnet 127.0.0.1 9042
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.
```

I have then edited my _Dockerfile_ to have both `cqlsh` ([installation instructions](https://docs.datastax.com/en/install/6.8/install/installCqlsh.html)) and `telnet` installed in the container for further tests:
```
FROM shotover/shotover-proxy:v0.1.5

RUN \
    apt-get update && \
    apt-get install -y python3 telnet

COPY files/cassandra_proxy/cqlsh-astra-20201104-bin.tar.gz /
RUN tar -xzvf cqlsh-astra-20201104-bin.tar.gz && rm cqlsh-astra-20201104-bin.tar.gz
RUN ln -s /cqlsh-astra/bin/cqlsh /bin/

RUN mkdir -m 755 /root/.cassandra
COPY files/cassandra_proxy/sf-class2-root.crt /root/.cassandra
COPY files/cassandra_proxy/cqlshrc  /root/.cassandra/cqlshrc

COPY files/cassandra_proxy/config.yaml /config/
COPY files/cassandra_proxy/topology.yaml /config/

COPY files/cassandra_proxy/sf-class2-root.crt /config/

ENTRYPOINT ["./shotover-proxy"]
```

where the _cqlshrc_ file has the configuration to use `cqlsh` with SSL and without the proxy for tests:
```
[connection]
port = 9142
factory = cqlshlib.ssl.ssl_transport_factory

[ssl]
validate = true
certfile =  /root/.cassandra/sf-class2-root.crt
version = TLSv1_2

[copy]
NUMPROCESSES=16
MAXATTEMPTS=25

[copy-from]
CHUNKSIZE=30
INGESTRATE=1500
MAXINSERTERRORS=-1
MAXPARSEERRORS=-1
MINBATCHSIZE=1
MAXBATCHSIZE=10

[csv]
field_size_limit=999999
```

I run and I exec in the container; from here I manage successfully to access AWS Keyspaces using `cqlsh` with SSL and without the proxy. The connection to the local port 9042 seems to work correctly here:
```
# telnet 127.0.0.1 9042
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
```

Eventually I deleted the _/root/.cassandra/cqlshrc_ file to avoid conflicts and I have tried to access AWS Keyspaces using `cqlsh` without SSL and with the proxy, but I get the same error that I get outside of the container:
```
Connection error: ('Unable to connect to any servers', {'127.0.0.1:9042': ConnectionShutdown('Connection to 127.0.0.1:9042 was closed')})
```

Nevertheless I have managed to achieve my goal by using a container with HAProxy instead of Shotover. If it is fine for you, I can close this ticket and share my solution with HAProxy in my closing comment.
